src/us/kbase/workspace/database/provenance/EventDate.java

```diff
@@ -0,0 +1,96 @@
+package us.kbase.workspace.database.provenance;
+
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+import us.kbase.workspace.database.Util;
+
+/**
+ * The date of a specified event.
+ */
+public class EventDate {
+
+	private final String date;
+	private final String dateFormat;
+	private final String event;
+
+	public static final Pattern VALID_DATE_FORMAT_REGEX = Pattern.compile("^YYYY(-MM(-DD)?)?$");
+
+	public static final Pattern VALID_DATE_REGEX = Pattern.compile("^[12]\\d{3}(-(0[1-9]|1[0-2])(-(0[1-9]|[12]\\d|3[01]))?)?$");
+
+	private EventDate(
+		final String date,
+		final String dateFormat,
+		final String event) {
+		this.date = date;
+		this.dateFormat = dateFormat;
+		this.event = event;
+	}
+
+	/** Get the date, for example "2022-05-10".
+	 * @return the date.
+	 */
+	public String getDate() {
+		return date;
+	}
+
+	/** Get the format of the date, for example "YYYY-MM-DD".
+	 * @return the date format.
+	 */
+	public String getDateFormat() {
+		return dateFormat;
+	}
+
+	/** Get the event that occurred on the date in question, for example "updated".
+	 * @return the event.
+	 */
+	public String getEvent() {
+		return event;
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(date, dateFormat, event);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		EventDate other = (EventDate) obj;
+		return Objects.equals(date, other.date) && Objects.equals(dateFormat, other.dateFormat)
+				&& Objects.equals(event, other.event);
+	}
+
+	/** Get a builder for an {@link EventDate}.
+	 * @return the builder.
+	 */
+	public static Builder getBuilder(final String date, final String dateFormat, final String event) {
+		return new Builder(date, dateFormat, event);
+	}
+
+	/** A builder for an {@link EventDate}. */
+	public static class Builder {
+
+		private String date;
+		private String dateFormat;
+		private String event;
+
+		private Builder(final String date, final String dateFormat, final String event) {
+			this.date = Common.checkAgainstRegex(date, VALID_DATE_REGEX, "date", false);
+			this.dateFormat = Common.checkAgainstRegex(dateFormat, VALID_DATE_FORMAT_REGEX, "dateFormat", false);
```

Comment:
It's possible for the date to not match the date format - that seems like it should be checked here, unless switching to a class instance for the date in which case the date format essentially acts like a sigfigs spec

Reply:
Crap, I meant to add a check for that but then forgot about it. :facepalm:

I was mulling over whether to use one of the date modules to parse the input, which would give better validation than just using a regex, and you could ensure that the `date` and `dateFormat` fields were consistent. Presumably I would have to use `Date`, rather than `Instant`, since we don't have millisecond (or in some cases, month!) precision for the dates to be captured.

Reply:
> better validation than just using a regex

that's where my head's at

>  Date, rather than Instant, since we don't have millisecond (or in some cases, month!) precision 

`Instant` should be fine with that. It can do anything `Date` can do (except be mutated). You shouldn't use `Date` if at all possible

Reply:
According to https://docs.oracle.com/javase/tutorial/datetime/iso/overview.html, `Instant` deals with (nano)seconds, and the constructor uses epoch time, so it's not going to be suitable for dates like `2022`, where only the year is available. The `date` field would need to be a `Year`, `LocalDate` (YMD), or possibly a `YearMonth`, to deal with the variations on the representation of the date that we have.

I am tempted to go back to storing the date as a string but validate it by attempting to convert it into either a `LocalDate`, a `YearMonth`, or a `Year`, and then normalise it by outputting it using a known `DateTimeFormatter` (e.g. `YYYY-MM`, `YYYY`, or `DateTimeFormatter.ISO_LOCAL_DATE`). 

Your thoughts?

Reply:
> constructor uses epoch time, so it's not going to be suitable for dates like 2022, where only the year is available. 

I'm not understanding why that is - you can certainly represent 2022 in epoch time. `Date` also uses epoch time under the hood.

```
final LocalDate ld = LocalDate.ofYearDay(2022, 1);
final Instant inst = ZonedDateTime.of(ld, LocalTime.MIN, ZoneId.of("Z")).toInstant();
System.out.println(inst);
```
prints
```
2022-01-01T00:00:00Z
```

Reply:
I think this is moot at this point